rubicon_ml/client/mixin.py

```diff
@@ -604,3 +604,102 @@ def tags(self) -> List[str]:
                 return self._domain.tags
 
         self._raise_rubicon_exception(return_err)
+
+
+class CommentMixin:
+    """Adds comment support to a client object."""
+
+    _domain: DOMAIN_TYPES
+
+    def _get_taggable_identifiers(self):
+        project_name, experiment_id = self._parent._get_identifiers()
+        entity_identifier = None
+
+        # experiments do not return an entity identifier - they are the entity
+        if isinstance(self, client.Experiment):
+            experiment_id = self.id
+        # dataframes and artifacts are identified by their `id`s
+        elif isinstance(self, client.Dataframe) or isinstance(self, client.Artifact):
+            entity_identifier = self.id
+        # everything else is identified by its `name`
+        else:
+            entity_identifier = self.name
+
+        return project_name, experiment_id, entity_identifier
+
+    @failsafe
+    def add_comments(self, comments: List[str]):
+        """Add comments to this client object.
+
+        Parameters
+        ----------
+        comments : list of str
+            The comment values to add.
+        """
+        if not isinstance(comments, list) or not all(
+            [isinstance(comment, str) for comment in comments]
+        ):
+            raise ValueError("`comments` must be `list` of type `str`")
+
+        project_name, experiment_id, entity_identifier = self._get_taggable_identifiers()
+
+        self._domain.add_comments(comments)
+        for repo in self.repositories:
+            repo.add_comments(
+                project_name,
+                comments,
+                experiment_id=experiment_id,
+                entity_identifier=entity_identifier,
+                entity_type=self.__class__.__name__,
+            )
+
+    @failsafe
+    def remove_comments(self, comments: List[str]):
+        """Remove comments from this client object.
+
+        Parameters
+        ----------
+        comments : list of str
+             The comment values to remove.
+        """
+        project_name, experiment_id, entity_identifier = self._get_taggable_identifiers()
+
+        self._domain.remove_comments(comments)
+        for repo in self.repositories:
+            repo.remove_comments(
```

Comment:
there is no `remove_comments` on the repo, is there? I think we said we weren't going to do removals for comments. at least not for now

Reply:
The way we implement `_update_tags` and `_update_comments` means that if we don't have the remove methods we end up adding two of each comment/tag. I figured adding `remove_comments` fixes that issue & and it doesn't hurt to have the option to remove comments in general even if it's not used much.

Reply:
I think it'd be better to update `_update_comments` and remove the references to deleting them. I'm not sure why it'd be adding multiples of each comment in its current state, so if that's occurring we should fix it without adding deletes

there is no `remove_comments` on the repository currently, so adding in deletes would add another chunk of work. I'd rather not expand the scope of this story at the moment